File: backend/app/schemas/metrics.py
"""Pydantic schemas for /api/v1/metrics/* endpoints (Phase 3)."""

from __future__ import annotations

from pydantic import BaseModel, Field


class PreviousPeriodOut(BaseModel):
    """Aggregate KPIs for the comparison period (same duration, shifted back).

    Returned alongside ``MetricsOverviewOut`` when ``compare_previous=True``
    is requested. Fields mirror the subset of ``MetricsOverviewOut`` that
    the frontend needs to compute deltas.
    """

    total_reviews: int
    avg_rating: float
    five_star_pct: float
    one_star_pct: float
    reply_rate_pct: float
    total_mentions: int
    period_start: str
    period_end: str


class MetricsOverviewOut(BaseModel):
    """Aggregate KPIs for the dashboard header cards."""

    total_reviews: int
    avg_rating: float
    five_star_pct: float
    one_star_pct: float
    total_with_comment: int
    total_with_reply: int
    reply_rate_pct: float = 0.0
    total_enotariado: int
    avg_rating_enotariado: float | None = None
    total_collaborators_active: int
    total_mentions: int
    rating_distribution: dict[str, int] = Field(
        default_factory=lambda: {"1": 0, "2": 0, "3": 0, "4": 0, "5": 0}
    )
    period_start: str
    period_end: str
    previous_period: PreviousPeriodOut | None = None


class MonthData(BaseModel):
    """Single month data point for the trends chart."""

    month: str
    total_reviews: int
    avg_rating: float
    reviews_enotariado: int
    avg_rating_enotariado: float | None = None
    reply_rate_pct: float = 0.0


class DataStatusOut(BaseModel):
    """Data freshness indicator for the dashboard chrome."""

    last_review_date: str | None
    last_collection_run: str | None
    total_reviews: int
    days_since_last_review: int | None


class TrendsOut(BaseModel):
    """Monthly trends over the selected period."""

    months: list[MonthData]


class CollaboratorMonthData(BaseModel):
    """Single month data point for a collaborator's mentions."""

    month: str
    mentions: int
    avg_rating: float | None = None


class CollaboratorMentionOut(BaseModel):
    """Aggregated mention statistics for one collaborator."""

    collaborator_id: int
    full_name: str
    is_active: bool
    total_mentions: int
    avg_rating_mentioned: float | None = None
    monthly: list[CollaboratorMonthData]


class CollaboratorMentionsOut(BaseModel):
    """Response wrapper for the collaborator mentions endpoint."""

    collaborators: list[CollaboratorMentionOut]


class MyPerformanceOut(BaseModel):
    """Personal performance metrics for a linked collaborator."""

    linked: bool
    collaborator_id: int | None = None
    full_name: str | None = None
    total_mentions: int = 0
    avg_rating: float | None = None
    ranking: int | None = None
    total_collaborators: int = 0
    monthly: list[CollaboratorMonthData] = []
    recent_reviews: list[dict] = []

File: frontend/src/types/metrics.ts
export type RatingBucket = '1' | '2' | '3' | '4' | '5'
export type RatingDistribution = Record<RatingBucket, number>

export interface PreviousPeriod {
  total_reviews: number
  avg_rating: number
  five_star_pct: number
  one_star_pct: number
  reply_rate_pct: number
  total_mentions: number
  period_start: string
  period_end: string
}

export interface MetricsOverview {
  total_reviews: number
  avg_rating: number
  five_star_pct: number
  one_star_pct: number
  total_with_comment: number
  total_with_reply: number
  reply_rate_pct: number
  total_enotariado: number
  avg_rating_enotariado: number | null
  total_collaborators_active: number
  total_mentions: number
  rating_distribution: RatingDistribution
  period_start: string
  period_end: string
  previous_period: PreviousPeriod | null
}

export type TrendsGranularity = 'month' | 'day'

export interface MonthData {
  month?: string
  day?: string
  total_reviews: number
  avg_rating: number
  reviews_enotariado: number
  avg_rating_enotariado: number | null
  reply_rate_pct: number
}

export interface DataStatus {
  last_review_date: string | null
  last_collection_run: string | null
  total_reviews: number
  days_since_last_review: number | null
}

export interface TrendsData {
  months: MonthData[]
  granularity?: TrendsGranularity
}

export interface CollaboratorMonthData {
  month: string
  mentions: number
  avg_rating: number | null
}

export interface CollaboratorMention {
  collaborator_id: number
  full_name: string
  is_active: boolean
  total_mentions: number
  avg_rating_mentioned: number | null
  monthly: CollaboratorMonthData[]
}

export interface CollaboratorMentionsData {
  collaborators: CollaboratorMention[]
}

export interface MyPerformance {
  linked: boolean
  collaborator_id: number | null
  full_name: string | null
  total_mentions: number
  avg_rating: number | null
  ranking: number | null
  total_collaborators: number
  monthly: CollaboratorMonthData[]
  recent_reviews: {
    review_id: string
    rating: number | null
    comment: string
    reviewer_name: string
    create_time: string | null
  }[]
}

export interface SystemUser {
  id: string
  email: string
  role: string
  is_active: boolean
}
